src/main/java/target/HttpTarget.java

```diff
@@ -62,68 +66,56 @@ public CompletableFuture<TargetResponse> call(final ConsumerRecord<String, Strin
+
+        var call = client.newCall(request);
+
+        var delay = Config.RETRY_POLICY_EXPONENTIAL_BACKOFF.get(0);
+        var maxDelay = Config.RETRY_POLICY_EXPONENTIAL_BACKOFF.get(1);
+        var delayFactor = Config.RETRY_POLICY_EXPONENTIAL_BACKOFF.get(2);
+        var maxDuration = Duration.ofMillis(Config.RETRY_POLICY_MAX_DURATION_MS);
+
+        var retryPolicy = RetryPolicy
+            .<Response>builder()
+            .withBackoff(delay, maxDelay, ChronoUnit.MILLIS, delayFactor)
+            .handleIf(e -> true)
+            .handleResultIf(r -> Integer.toString(r.code()).matches(Config.RETRY_PROCESS_WHEN_STATUS_CODE_MATCH))
+            .withMaxDuration(maxDuration)
+            .build();
+
+        return FailsafeCall
+            .with(retryPolicy)
+            .compose(call)
+            .executeAsync()
+            .handleAsync(
+                (response, throwable) -> {
+                    if (throwable != null && Config.RETRY_TOPIC != null) {
+                        return producer.produce(Config.RETRY_TOPIC, record);
                     }
-                );
 
-            var result = new CompletableFuture<Optional<HttpResponse<String>>>();
+                    var statusCode = Integer.toString(response.code());
 
-            httpFuture.whenComplete(
-                (ok, error) -> {
-                    if (error != null) {
-                        result.completeExceptionally(error);
-                    } else {
-                        result.complete(Optional.of(ok));
+                    if (statusCode.matches(Config.PRODUCE_TO_RETRY_TOPIC_WHEN_STATUS_CODE_MATCH)) {
+                        Monitor.processMessageError();
+                        if (Config.RETRY_TOPIC != null) {
+                            producer.produce(Config.RETRY_TOPIC, record);
+                            Monitor.retryProduced(record);
+                            return;
+                        }
                     }
-                }
-            );
 
-            return result;
-        };
-
-        Optional<HttpResponse<String>> defaultResponse = Optional.empty();
-
-        return Failsafe
-            .with(
-                Fallback.of(defaultResponse),
-                retryPolicy.get(record, o -> o.map(r -> r.statusCode()).orElseGet(() -> -1))
-            )
-            .getStageAsync(completionStageCheckedSupplier)
-            .thenApplyAsync(
-                optionalResponse -> {
-                    if (optionalResponse.isEmpty()) {
-                        return new TargetResponse(OptionalLong.empty(), OptionalLong.empty());
+                    if (statusCode.matches(Config.PRODUCE_TO_DEAD_LETTER_TOPIC_WHEN_STATUS_CODE_MATCH)) {
+                        Monitor.processMessageError();
+                        if (Config.DEAD_LETTER_TOPIC != null) {
+                            producer.produce(Config.DEAD_LETTER_TOPIC, record);
+                            Monitor.deadLetterProcdued(record);
```

Comment:
maybe its better to put the Monitor inside the producer itself

Reply:
something like `Monitor.produced(topicName)`, that can be in the inner producer call (when the Future resolves).
This can also be here, but only when the Future is completed and not the way its currently written

Reply:
Yep that what I did